sql/lantern.sql

```diff
@@ -193,3 +195,223 @@ BEGIN
+RETURNS NAME AS $$
+DECLARE
+  stmt TEXT;
+  res REAL[];
+  codebooks REAL[][][];
+  i INT;
+  subset_len INT;
+  end_idx INT;
+  codebook_table NAME;
+  dim INT;
+BEGIN
+  
+  stmt := format('SELECT array_length(%I, 1) FROM %I WHERE %1$I IS NOT NULL LIMIT 1', col, tbl);
+  EXECUTE stmt INTO dim;
+
+	-- Get codebooks
+	codebooks := _lantern_internal.create_pq_codebook(tbl, col, cluster_cnt, subset_count, distance_metric);
+
+	-- Create codebook table
+  codebook_table := format('_lantern_codebook_%s', tbl);
+  stmt := format('DROP TABLE IF EXISTS %s CASCADE', codebook_table);
+  EXECUTE stmt;
+  
+  stmt:= format('CREATE TABLE %s(subvector_id INT, centroid_id INT, c REAL[]);', codebook_table);
+  EXECUTE stmt;
+  
+  stmt:= format('CREATE INDEX ON %s USING BTREE(subvector_id, centroid_id);', codebook_table);
+  EXECUTE stmt;
+  
+  -- Iterate over codebooks and insert into table
+  FOR i IN 1..subset_count loop
+  	FOR k IN 1..cluster_cnt loop
+  	  -- centroid_id is k-1 because k is in range[0,255] but postgres arrays start from index 1
+      stmt := format('INSERT INTO %I(subvector_id, centroid_id, c) VALUES (%s, %s, ARRAY(SELECT * FROM unnest(''%s''::REAL[])))', codebook_table, i - 1, k - 1, codebooks[i:i][k:k]);
+      EXECUTE stmt;
+  	END LOOP;
+  END LOOP;
+
+  return codebook_table;
+END;
+$$ LANGUAGE plpgsql;
+
+-- Compress vector using codebook
+CREATE OR REPLACE FUNCTION _lantern_internal.compress_vector(v REAL[], subset_count INTEGER, codebook regclass, distance_metric TEXT)
+RETURNS pqvec AS $$
+DECLARE
+  subset_center INT;
+  start_idx INT;
+  end_idx INT;
+  dim INT;
+  subset_len INT;
+  res INT[];
+  subvector_id INT;
+BEGIN
+  dim := array_length(v, 1);
+  res := '{}'::INT[];
+  subset_len := dim/subset_count;
+  subvector_id := 0;
+
+  FOR i IN 1..dim BY subset_len LOOP
+    IF i = dim THEN
+      end_idx := dim;
+    ELSE
+      end_idx := i + subset_len - 1;
+    END IF;
+    EXECUTE format('SELECT centroid_id FROM %I WHERE subvector_id=%s ORDER BY %s_dist(c, %L) LIMIT 1', codebook, subvector_id, distance_metric, v[i:end_idx]) INTO subset_center;
+    res := array_append(res, subset_center);
+    subvector_id := subvector_id + 1;
+  END LOOP;
+  
+  RETURN res::pqvec;
+END;
+$$ LANGUAGE plpgsql IMMUTABLE;
+
+CREATE OR REPLACE FUNCTION compress_vector(v REAL[], codebook regclass, distance_metric TEXT)
+RETURNS pqvec AS $$
+DECLARE
+  subset_count INT;
+  stmt TEXT;
+BEGIN
+
+  stmt := format('SELECT COUNT(centroid_id) FROM %I WHERE centroid_id=1', codebook);
+  EXECUTE stmt INTO subset_count;
+  RETURN _lantern_internal.compress_vector(v, subset_count, codebook, distance_metric);
+END;
+$$ LANGUAGE plpgsql;
+
+-- Decompress vector using codebook
+CREATE OR REPLACE FUNCTION decompress_vector(v pqvec, codebook regclass)
+RETURNS REAL[] AS $$
+DECLARE
+  res REAL[];
+  subset REAL[];
+  centroid_id INT;
+  subvector_id INT;
+BEGIN
+  res := '{}'::REAL[];
+  subvector_id := 0;
+  FOREACH centroid_id in array v::INT[]
```

Comment:
This is a public function, so worth doing more thorough argument validation and throw helpful error messages early
e.g. codebook may be a relation that is not a codebook, `v` might be empty or have the wrong size

if `v` is empty, currently this function returns the cryptic `ERROR: FOREACH expression must not be null`, which is not ideal, as the caller has no idea we use FOREACH in here